packages/astro/src/core/render-context.ts

```diff
@@ -470,12 +497,13 @@ export class RenderContext {
 				propagators: new Set(),
 			},
 			shouldInjectCspMetaTags: !!manifest.csp,
-			scriptHashes: manifest.csp?.scriptHashes ?? [],
-			scriptResources: manifest.csp?.scriptResources ?? [],
-			styleHashes: manifest.csp?.styleHashes ?? [],
-			styleResources: manifest.csp?.styleResources ?? [],
 			cspAlgorithm: manifest.csp?.algorithm ?? 'SHA-256',
-			directives: manifest.csp?.directives ?? [],
+			// The following arrays must be cloned, otherwise they become mutable across routes.
+			scriptHashes: !!manifest.csp?.scriptHashes ? [...manifest.csp.scriptHashes] : [],
+			scriptResources: !!manifest.csp?.scriptResources ? [...manifest.csp.scriptResources] : [],
+			styleHashes: !!manifest.csp?.styleHashes ? [...manifest.csp.styleHashes] : [],
+			styleResources: !!manifest.csp?.styleResources ? [...manifest.csp.styleResources] : [],
+			directives: !!manifest.csp?.directives ? [...manifest.csp.directives] : [],
```

Comment:
```suggestion
			scriptHashes: manifest.csp?.scriptHashes ? [...manifest.csp.scriptHashes] : [],
			scriptResources: manifest.csp?.scriptResources ? [...manifest.csp.scriptResources] : [],
			styleHashes: manifest.csp?.styleHashes ? [...manifest.csp.styleHashes] : [],
			styleResources: manifest.csp?.styleResources ? [...manifest.csp.styleResources] : [],
			directives: manifest.csp?.directives ? [...manifest.csp.directives] : [],
```
This is a conditional so no need to coerce here